assets/js/panel.js

```diff
@@ -0,0 +1,197 @@
+      .classed("closed", false);
+
+    // disable the info button; inactive when the intro is visible
+    this.infoButton.disabled = true;
+
+    // transfer focus to the panel
+    this.container.querySelector("#panel").focus();
+
+    // fixme: doesn't work ?
+    document.body.dataset.panelvisible = true;
+  }
+
+  get detailsVisible() {
+    return this.container.classList.contains("show-details");
+  }
+
+  close(closeDetails = true) {
+    // close the intro and enable the info button
+    // by default, closes panel entireuly
+
+    // determine if leaf details are currently displayed
+    // use dataset.showing ?
+    let leafVisible = this.detailsVisible;
+
+    // if leaf details are visible and close details is true,
+    // deselect the leaf currently displayed, close that also,
+    // unless closeDetails has been disabled;
+    // (has a side effect of also removing any currently selected tag)
+    if (leafVisible && closeDetails) {
+      this.container.classList.remove("show-details");
+      // clear out stored value for loaded url
+      delete this.el.dataset.showing;
+      this.el.dispatchEvent(PanelCloseEvent);
+    }
+
+    // if we are closing everything or no leaf is visible, close the panel
+    if (closeDetails || !leafVisible) {
+      this.container.classList.add("closed");
+      // update attribute on body to control overflow behavior
+      document.body.dataset.panelvisible = false;
+    }
+
+    // enable the info button;
+    // provides a way to get back to the intro on mobile
+    this.infoButton.disabled = false;
+  }
+
+  showIntro() {
+    // dispatch panel-close event;
+    // handler in leaf code will close leaf details
+    // (showing the intro implies closing leaf details if a leaf is selected)
+    this.el.dispatchEvent(PanelCloseEvent);
+    // open the panel without showing leaf details section
+    this.open(false);
+  }
+
+  closeIntro() {
+    // close the panel without closing leaf details
+    this.close(false);
+  }
+
+  announce(content) {
+    // update the contents of the aria live region
+    // to make polite announcements for screen reader users
+    this.status.textContent = content;
+  }
+
+  loadURL(url, callback) {
+    // load specified url; display article contents in the panel
+    // takes an optional callback; if defined, will be applied to
+    // the url contents before inserting
+
+    // if the requested url is already showing, do nothing
+    if (this.el.dataset.showing && this.el.dataset.showing == url) {
+      return;
+    }
+    // If you need to test load failure, uncomment this
+    // if (Math.random() < 0.5) {
+    //   url = url + "xxx";
+    // }
+    // console.log("fetching:", url);
+
+    fetch(url)
+      .then((response) => {
+        if (!response.ok) {
+          return Promise.reject(response);
+        }
+        return response;
+      })
+      .then((response) => response.text())
+      .then((html) => {
+        let parser = new DOMParser();
```

Comment:
If I understand correctly, there are pre-generated HTML articles ('leaves'?) that can be either viewed as separate pages, or some of their content gets embedded here. This is a bit unusual, and I would prefer a data-driven approach.

The data for each leaf / article could be stored as markdown with a yaml header (or just JSON if it's simpler) and should be used to generate the standalone pages, as well as being called here and displayed through a template. This is another example of where a framework would come in handy.

Evaluating the HTML via DOMParser may also be a little unsafe? Though I'm not sure how the fetch call would be intercepted, but it's still an opening. I think ultimately fornt-end frameworks perform equivalent operations, but they have likely thought out security issues. For example React handles changes by modifying a "virtual" DOM before pushing changes to the actual page DOM.